RxJava1ContextPropagator does not work correctly in the dev mode

Quarkus build: https://github.com/quarkusio/quarkus/commit/ef71554148994a2e43e619dc4f5e2b3f692be5eb

# Stack trace:
```
Caused by: java.lang.IllegalStateException: Container not running: ArcContainerImpl [id=ff6e4c3c-6b02-424a-8290-37430a072755, running=false, beans=0, observers=0, scopes=[interface javax.inject.Singleton, interface javax.enterprise.context.ApplicationScoped]]
	at io.quarkus.arc.ArcContainerImpl.requireRunning(ArcContainerImpl.java:548)
	at io.quarkus.arc.ArcContainerImpl.requestContext(ArcContainerImpl.java:239)
	at io.quarkus.arc.runtime.context.ArcContextProvider.lambda$currentContext$4(ArcContextProvider.java:42)
	at io.smallrye.context.impl.ActiveContextState.<init>(ActiveContextState.java:18)
	at io.smallrye.context.impl.CapturedContextState.begin(CapturedContextState.java:35)
	at io.smallrye.context.SmallRyeThreadContext.lambda$withContext$0(SmallRyeThreadContext.java:213)
	at io.smallrye.context.propagators.rxjava1.ContextPropagatorOnObservableCreateAction$ContextCapturerObservable$OnAssemblyObservableSubscriber.onNext(ContextPropagatorOnObservableCreateAction.java:62)
	at rx.internal.operators.NotificationLite.accept(NotificationLite.java:135)
```

# Steps to reproduce:
1. clone https://github.com/mkouba/pulls
2. run `mvn compile quarkus:dev`
3. hit `http://localhost:8080/pulls`
4. make some irrelevant change
5. hit `http://localhost:8080/pulls`

It seems that SM CP is using some outdated `ThreadContextSnapshot` returned from `ArcContextProvider#currentContext()` - this snapshot is referencing the stopped ArC container instance. A quick solution might be to use `Arc.container()` in the returned lambda.

--- a/core/runtime/src/main/java/io/quarkus/runtime/CleanableExecutor.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/CleanableExecutor.java
@@ -1,0 +1,254 @@
+package io.quarkus.runtime;
+
+import java.lang.reflect.Field;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.jboss.threads.EnhancedQueueExecutor;
+
+import sun.misc.Unsafe;
+
+/**
+ * An executor that can be 'cleaned'. Each clean increments an internal generation number. When a task is executed by
+ * a thread that has a different generation number to the last task executed by the thread then the threads ThreadLocal's
+ * will be cleared. This means that thread locals from a previous deployment cannot interfere with the current deployment.
+ *
+ * This is only for development mode, it must not be used for production applications.
+ *
+ * TODO: should this just provide a facacde that simply starts a new thread pool instead?
+ */
+public final class CleanableExecutor implements ExecutorService {
+
+    private final EnhancedQueueExecutor executor;
+
+    private static final AtomicInteger generation = new AtomicInteger(1);
+    private final ThreadLocal<Integer> lastGeneration = new ThreadLocal<Integer>() {
+        @Override
+        protected Integer initialValue() {
+            return -1;
+        }
+    };
+
+    public CleanableExecutor(EnhancedQueueExecutor executor) {
+        this.executor = executor;
+    }
+
+    public void clean() {
+        //also clear the current thread, as this is called by the hot deployment thread so it is about to hit a new deployment
+        Resetter.run();
+        generation.incrementAndGet();
+        CountDownLatch latch = new CountDownLatch(1);
+        Runnable empty = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        };
+        //this is a hack, basically just submit a heap of tasks to the pool in order to try and clean up all
+        //threadlocal state. It does not really matter if this does not work, however if these threads are holding
+        //state that is stopping things being GC'ed it can help with memory usage
+        try {
+            for (int i = 0; i < executor.getMaximumPoolSize(); ++i) {
+                submit(empty);
+            }
+        } finally {
+            latch.countDown();
+        }
+
+    }
+
+    private void handleClean(int taskGen) {
+        int val = lastGeneration.get();
+        if (val == -1) {
+            lastGeneration.set(taskGen);
+        } else if (val != taskGen) {
+            Resetter.run();
+            lastGeneration.set(taskGen);
+        }
+    }
+
+    @Override
+    public void shutdown() {
+        executor.shutdown();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
+        return executor.shutdownNow();
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return executor.isShutdown();
+    }
+
+    @Override
+    public boolean isTerminated() {
+        return executor.isTerminated();
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        return executor.awaitTermination(timeout, unit);
+    }
+
+    @Override
+    public <T> Future<T> submit(Callable<T> task) {
+        return executor.submit(new CleaningCallable<>(task));
+    }
+
+    @Override
+    public <T> Future<T> submit(Runnable task, T result) {
+        return executor.submit(new CleaningRunnable(task), result);
+    }
+
+    @Override
+    public Future<?> submit(Runnable task) {
+        return executor.submit(new CleaningRunnable(task));
+    }
+
+    @Override
+    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) throws InterruptedException {
+        ArrayList<Callable<T>> submit = createWrappedList(tasks);
+        return executor.invokeAll(submit);
+    }
+
+    @Override
+    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit)
+            throws InterruptedException {
+        ArrayList<Callable<T>> submit = createWrappedList(tasks);
+        return executor.invokeAll(submit, timeout, unit);
+    }
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks) throws InterruptedException, ExecutionException {
+        ArrayList<Callable<T>> submit = createWrappedList(tasks);
+        return executor.invokeAny(submit);
+    }
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        ArrayList<Callable<T>> submit = createWrappedList(tasks);
+        return executor.invokeAny(submit, timeout, unit);
+    }
+
+    private <T> ArrayList<Callable<T>> createWrappedList(Collection<? extends Callable<T>> tasks) {
+        ArrayList<Callable<T>> submit = new ArrayList<>();
+        for (Callable<T> i : tasks) {
+            submit.add(new CleaningCallable<>(i));
+        }
+        return submit;
+    }
+
+    @Override
+    public void execute(Runnable command) {
+        executor.submit(new CleaningRunnable(command));
+    }
+
+    /**
+     * copied from JBoss threads
+     */
+    static final class Resetter {
+        private static final long threadLocalMapOffs;
+        private static final long inheritableThreadLocalMapOffs;
+
+        static {
+            final Field threadLocals = AccessController.doPrivileged(new DeclaredFieldAction(Thread.class, "threadLocals"));
+            threadLocalMapOffs = threadLocals == null ? 0 : unsafe.objectFieldOffset(threadLocals);
+            final Field inheritableThreadLocals = AccessController
+                    .doPrivileged(new DeclaredFieldAction(Thread.class, "inheritableThreadLocals"));
+            inheritableThreadLocalMapOffs = inheritableThreadLocals == null ? 0
+                    : unsafe.objectFieldOffset(inheritableThreadLocals);
+        }
+
+        static void run() {
+            final Thread thread = Thread.currentThread();
+            if (threadLocalMapOffs != 0)
+                unsafe.putObject(thread, threadLocalMapOffs, null);
+            if (inheritableThreadLocalMapOffs != 0)
+                unsafe.putObject(thread, inheritableThreadLocalMapOffs, null);
+        }
+    }
+
+    static final Unsafe unsafe;
+
+    static {
+        unsafe = AccessController.doPrivileged(new PrivilegedAction<Unsafe>() {
+            public Unsafe run() {
+                try {
+                    final Field field = Unsafe.class.getDeclaredField("theUnsafe");
+                    field.setAccessible(true);
+                    return (Unsafe) field.get(null);
+                } catch (IllegalAccessException e) {
+                    throw new IllegalAccessError(e.getMessage());
+                } catch (NoSuchFieldException e) {
+                    throw new NoSuchFieldError(e.getMessage());
+                }
+            }
+        });
+    }
+
+    static final class DeclaredFieldAction implements PrivilegedAction<Field> {
+        private final Class<?> clazz;
+        private final String fieldName;
+
+        DeclaredFieldAction(final Class<?> clazz, final String fieldName) {
+            this.clazz = clazz;
+            this.fieldName = fieldName;
+        }
+
+        public Field run() {
+            try {
+                return clazz.getDeclaredField(fieldName);
+            } catch (NoSuchFieldException e) {
+                return null;
+            }
+        }
+    }
+
+    private class CleaningRunnable implements Runnable {
+        private final Runnable command;
+        final int gen = generation.get();
+
+        public CleaningRunnable(Runnable command) {
+            this.command = command;
+        }
+
+        @Override
+        public void run() {
+            handleClean(gen);
+            command.run();
+        }
+    }
+
+    private class CleaningCallable<T> implements Callable<T> {
+        private final Callable<T> i;
+        final int gen = generation.get();
+
+        public CleaningCallable(Callable<T> i) {
+            this.i = i;
+        }
+
+        @Override
+        public T call() throws Exception {
+            handleClean(gen);
+            return i.call();
+        }
+    }
+}

--- a/core/runtime/src/main/java/io/quarkus/runtime/ExecutorTemplate.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/ExecutorTemplate.java
@@ -45,31 +45,39 @@ public class ExecutorTemplate {
      * In dev mode for now we need the executor to last for the life of the app, as it is used by Undertow. This will likely
      * change
      */
-    static ExecutorService devModeExecutor;
+    static CleanableExecutor devModeExecutor;
 
     public ExecutorService setupRunTime(ShutdownContext shutdownContext, ThreadPoolConfig threadPoolConfig,
             LaunchMode launchMode) {
         if (devModeExecutor != null) {
             return devModeExecutor;
         }
-        final EnhancedQueueExecutor executor = createExecutor(threadPoolConfig);
-
-        Runnable shutdownTask = createShutdownTask(threadPoolConfig, executor);
+        final EnhancedQueueExecutor underlying = createExecutor(threadPoolConfig);
+        ExecutorService executor;
+        Runnable shutdownTask = createShutdownTask(threadPoolConfig, underlying);
         if (launchMode == LaunchMode.DEVELOPMENT) {
-            devModeExecutor = executor;
+            devModeExecutor = new CleanableExecutor(underlying);
+            shutdownContext.addShutdownTask(new Runnable() {
+                @Override
+                public void run() {
+                    devModeExecutor.clean();
+                }
+            });
+            executor = devModeExecutor;
             Runtime.getRuntime().addShutdownHook(new Thread(shutdownTask, "Executor shutdown thread"));
         } else {
             shutdownContext.addShutdownTask(shutdownTask);
+            executor = underlying;
         }
         return executor;
     }
 
     public static ExecutorService createDevModeExecutorForFailedStart(ThreadPoolConfig config) {
-        EnhancedQueueExecutor executor = createExecutor(config);
-        Runnable task = createShutdownTask(config, executor);
-        devModeExecutor = executor;
+        EnhancedQueueExecutor underlying = createExecutor(config);
+        Runnable task = createShutdownTask(config, underlying);
+        devModeExecutor = new CleanableExecutor(underlying);
         Runtime.getRuntime().addShutdownHook(new Thread(task, "Executor shutdown thread"));
-        return executor;
+        return devModeExecutor;
     }
 
     private static Runnable createShutdownTask(ThreadPoolConfig threadPoolConfig, EnhancedQueueExecutor executor) {
@@ -163,4 +171,5 @@ public class ExecutorTemplate {
         builder.setKeepAliveTime(threadPoolConfig.keepAliveTime);
         return builder.build();
     }
+
 }

--- a/extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java
+++ b/extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java
@@ -43,13 +43,17 @@ import io.quarkus.arc.processor.AnnotationsTransformer;
 import io.quarkus.arc.processor.BuiltinScope;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.ExecutionTime;
+import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.substrate.SubstrateSystemPropertyBuildItem;
 import io.quarkus.deployment.logging.LogCleanupFilterBuildItem;
 import io.quarkus.smallrye.faulttolerance.runtime.QuarkusFallbackHandlerProvider;
 import io.quarkus.smallrye.faulttolerance.runtime.QuarkusFaultToleranceOperationProvider;
+import io.quarkus.smallrye.faulttolerance.runtime.SmallryeFaultToleranceTemplate;
 import io.smallrye.faulttolerance.DefaultCommandListenersProvider;
 import io.smallrye.faulttolerance.DefaultHystrixConcurrencyStrategy;
 import io.smallrye.faulttolerance.HystrixCommandBinding;
@@ -160,5 +164,11 @@ public class SmallRyeFaultToleranceProcessor {
         logCleanupFilter.produce(new LogCleanupFilterBuildItem("com.netflix.config.sources.URLConfigurationSource",
                 "No URLs will be polled as dynamic configuration sources.",
                 "To enable URLs as dynamic configuration sources"));
+    }
+
+    @Record(ExecutionTime.STATIC_INIT)
+    @BuildStep
+    public void clearStatic(SmallryeFaultToleranceTemplate template, ShutdownContextBuildItem context) {
+        template.resetCommandContextOnUndeploy(context);
     }
 }

--- a/extensions/smallrye-fault-tolerance/runtime/src/main/java/io/quarkus/smallrye/faulttolerance/runtime/SmallryeFaultToleranceTemplate.java
+++ b/extensions/smallrye-fault-tolerance/runtime/src/main/java/io/quarkus/smallrye/faulttolerance/runtime/SmallryeFaultToleranceTemplate.java
@@ -1,0 +1,30 @@
+package io.quarkus.smallrye.faulttolerance.runtime;
+
+import com.netflix.hystrix.Hystrix;
+import com.netflix.hystrix.metric.HystrixCollapserEventStream;
+import com.netflix.hystrix.metric.HystrixCommandCompletionStream;
+import com.netflix.hystrix.metric.HystrixCommandStartStream;
+import com.netflix.hystrix.metric.HystrixThreadEventStream;
+import com.netflix.hystrix.metric.HystrixThreadPoolCompletionStream;
+
+import io.quarkus.runtime.ShutdownContext;
+import io.quarkus.runtime.annotations.Template;
+
+@Template
+public class SmallryeFaultToleranceTemplate {
+
+    public void resetCommandContextOnUndeploy(ShutdownContext context) {
+        context.addShutdownTask(new Runnable() {
+            @Override
+            public void run() {
+                HystrixCommandCompletionStream.reset();
+                HystrixCollapserEventStream.reset();
+                HystrixCommandStartStream.reset();
+                HystrixThreadPoolCompletionStream.reset();
+                HystrixCommandStartStream.reset();
+                HystrixThreadEventStream.getInstance().shutdown();
+                Hystrix.reset();
+            }
+        });
+    }
+}

